cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -233,6 +235,184 @@ struct DivideChecked {
   }
 };
 
+struct Power {
+  template <typename T, typename Arg0, typename Arg1>
+  static enable_if_signed_integer<T> Call(KernelContext* ctx, Arg0 left, Arg1 right) {
+    if (right < 0) {
+      ctx->SetStatus(
+          Status::Invalid("integers to negative integer powers are not allowed"));
+    }
+    if (left == 0 && right != 0) {
+      return 0;
+    }
+    Arg0 result = 1;
+    for (Arg1 i = 0; i < right; i++) {
+      MultiplyWithOverflow(result, left, &result);
+    }
+    return result;
+  }
+
+  template <typename T, typename Arg0, typename Arg1>
+  static enable_if_unsigned_integer<T> Call(KernelContext* ctx, Arg0 left, Arg1 right) {
+    if (left == 0 && right != 0) {
+      return 0;
+    }
+    Arg0 result = 1;
+    for (Arg1 i = 0; i < right; i++) {
+      MultiplyWithOverflow(result, left, &result);
+    }
+    return result;
+  }
+
+  template <typename T, typename Arg0, typename Arg1>
+  static enable_if_floating_point<T> Call(KernelContext* ctx, Arg0 left, Arg1 right) {
+    if (std::isnan(left) || std::isnan(right)) {
+      return NAN;
+    }
```

Comment:
It can, but it will return `pow(1, NaN) -> 1` and `pow(NaN, 0) -> 1` while we want to return `pow(1, NaN) -> NaN` and `pow(NaN, 0) -> NaN` in the default case.

Reply:
NumPy returns 1 by default in those cases as well:

```
In [20]: np.power(1.0, np.nan)
Out[20]: 1.0

In [21]: np.power(np.nan, 0.0)
Out[21]: 1.0
```

So we could maybe do that as well? (for NaN, not necessarily for Null)

Reply:
Ok I changed the option from `remove_nulls` to `propagate_nulls` and it is set to `false` by default. So default behavior will match numpy.

Reply:
Don't mix `NaN` with `Null`.
`NaN` here is a floating number called `Not a Number`. It is a useful and valid floating point value.
`Null` means the data point is missing.

Reply:
I think for NaN we can simply always follow the above behaviour like numpy (not controlled by any option).